File: src/autonomous-trust/autonomous_trust/core/_python/freshness.py
# ******************
#  Copyright 2026 TekFive, Inc., Sean M. Brennan, and contributors
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
# ******************
"""Per-verb message freshness: a monotonic sender sequence plus a receiver-side
high-water mark.

A signed AT message binds its content and nothing about when it was said (see
``Message._signable_content``), so a verb whose payload carries no freshness
token can be captured and re-presented, to its own recipient or to a different
one. Verbs that already carry such a token -- a Paxos ballot id, a group key
epoch, a chain index, an attestation nonce, a relayed query id -- are unaffected
and must NOT be given a second one. This module is for the verbs that carried
nothing. See ``doc/architecture/security-hardening.md``, "Replay resistance,
per verb".

The shape, and why this one:

* The SENDER stamps a monotonically increasing ``seq`` into the payload, which
  the message signature already covers because the pre-image includes
  ``base64(data)``. So a stripped or edited ``seq`` is an invalid message
  rather than an unstamped one -- no separate integrity story is needed, and no
  change to the pre-image, which is what would have made this a wire break for
  the C twin and the byte-pinned corpus.
* The RECEIVER keeps, per (sender, verb), the highest ``seq`` it has accepted,
  and refuses anything at or below it. Not a timestamp: a clock-based window
  would make the deliberately advisory cohort-clock subsystem
  (``network/clock.py``, ``doc/architecture/cohort-clock-skew.md``)
  load-bearing, and would still leave replay free inside the window.
* ONE counter per process, shared across the verbs that process emits, rather
  than one counter per verb. The marks are keyed per (sender, verb), so a
  shared counter is still strictly increasing within any single verb -- the
  gaps a verb sees are just the other verbs' sends. One counter means one
  number to persist and one place for it to go wrong.

Both halves are persisted, and both have to be. A sender that rewound its
counter after a restart would have its next messages refused as replays by
peers whose marks it cannot see -- the same trap the checkpoint and slash epoch
counters document. A receiver that forgot its marks would accept one replay per
(sender, verb) on every restart, which for an admission verb is one
re-admission of a peer that was removed.

There is no lenient mode for an unstamped message. That is the standing rule
for this kind of change here (``doc/architecture/reputation.md``, "Quorum
attestation"): a receiver that accepts unstamped messages is a receiver an
attacker selects by simply not stamping.
"""

import json
import os

from .config import Configuration, atomic_write

#: Base name of the per-process freshness file; the process name is appended so
#: two processes on one node never write the same file, and so a counter is
#: only ever advanced by the single process that owns it.
FRESHNESS_FILE = 'freshness'

#: Coalescing window for mark writes, seconds. A mark advance is one small
#: atomic write, and on a busy verb that is per accepted message; bursts inside
#: this window are written once. The residual is bounded and stated: an unclean
#: stop can lose up to this much of the mark state, which costs at most one
#: replay per (sender, verb) whose mark had not yet reached disk. The sender
#: counter is NOT throttled -- it is written before the message it stamps goes
#: out, because a rewound counter breaks liveness rather than merely narrowing
#: a window.
FLUSH_INTERVAL_SECS = float(os.environ.get('AT_FRESHNESS_FLUSH_SECS', '1.0'))


class Freshness(object):
    """One process's freshness state: its own send counter, and its marks.

    ``clock`` is injected (defaulting to the module's monotonic source) so a
    test can drive the flush window without sleeping.
    """

    def __init__(self, proc_name: str, logger=None, clock=None):
        self.proc_name = proc_name
        self.logger = logger
        self._clock = clock if clock is not None else _monotonic
        self._seq = 0
        self._marks: dict[tuple, int] = {}
        # Per-verb count of messages this process REFUSED as stale. Not
        # persisted and not part of the guard's decision: the marks are the
        # security state, this is only the evidence that they fired. A refusal
        # is otherwise invisible from outside — the pre-existing per-sender
        # cooldowns suppress a second delivery on their own, so "no second
        # response" cannot tell the mark from the cooldown. Read by the
        # conformance `freshness_refusals` observable; C mirrors it in
        # freshness_refusals().
        self._refusals: dict[str, int] = {}
        self._dirty = False
        self._last_flush = 0.0
        # Resolved ONCE, here: the config root is established before the
        # process is built (the conformance harness sets it per scenario), and
        # re-resolving on every write would let the file move under a running
        # process -- which for replay marks means silently starting over.
        self._path = os.path.join(
            Configuration.get_cfg_dir(),
            '%s-%s%s' % (FRESHNESS_FILE, self.proc_name,
                         Configuration.file_ext))
        self._load()

    # --- paths ---------------------------------------------------------

    def path(self):
        return self._path

    # --- sender side ---------------------------------------------------

    def stamp(self) -> int:
        """The next sequence number to put on an outgoing payload.

        Persisted before it is returned: the caller is about to put this number
        on the wire, and a number that reaches a peer but not the disk is
        exactly the one that gets reused after a restart and refused.
        """
        self._seq += 1
        self._flush(force=True)
        return self._seq

    # --- receiver side -------------------------------------------------

    def accept(self, sender, verb, seq) -> bool:
        """True iff ``seq`` is fresh for this (sender, verb), advancing the
        mark when it is.

        A missing, non-integer or non-positive ``seq`` is refused: sequences
        start at 1, so 0 is the never-seen floor and anything that cannot be
        read as a number above it is either an unstamped sender or a stripped
        field, both of which are refusals rather than exceptions.
        """
        if sender is None or seq is None:
            self._refused(verb)
            return False
        try:
            seq = int(seq)
        except (TypeError, ValueError):
            self._refused(verb)
            return False
        key = (str(sender), str(verb))
        if seq <= self._marks.get(key, 0):
            self._refused(verb)
            return False
        self._marks[key] = seq
        self._dirty = True
        self._flush()
        return True

    def mark(self, sender, verb) -> int:
        """The current high-water mark, for logging and tests."""
        return self._marks.get((str(sender), str(verb)), 0)

    def _refused(self, verb) -> None:
        """Tally one refusal against ``verb``.

        Keyed by verb alone, not by (sender, verb) like the marks: the sender
        detail is already in the caller's log line and its ``_probes`` counter,
        and a per-verb total is what an observer outside the process can
        assert. Both refusal shapes land here — a sequence at or below the mark
        (a replay) and one that is missing, unreadable or non-positive (an
        unstamped sender, or a stripped field) — because to the guard they are
        one decision. Which of the two a case induces is stated by the case.
        """
        key = str(verb)
        self._refusals[key] = self._refusals.get(key, 0) + 1

    def refusals(self, verb=None) -> int:
        """Messages refused as stale, for ``verb`` or across every verb."""
        if verb is None:
            return sum(self._refusals.values())
        return self._refusals.get(str(verb), 0)

    # --- persistence ---------------------------------------------------

    def _flush(self, force=False):
        if not force and not self._dirty:
            return
        nowish = self._clock()
        if not force and (nowish - self._last_flush) < FLUSH_INTERVAL_SECS:
            return
        doc = {
            'seq': int(self._seq),
            'marks': {'%s|%s' % (s, v): int(n)
                      for (s, v), n in self._marks.items()},
        }
        try:
            with atomic_write(self.path()) as f:
                json.dump(doc, f, indent=2)
            self._dirty = False
            self._last_flush = nowish
        except (OSError, IOError, ValueError, TypeError) as e:
            # Swallowed, like the other snapshot writers: the alternative is
            # taking a process down over a full disk. Logged at warning rather
            # than debug because the cost is a reopened replay window, not a
            # slower warm start.
            if self.logger is not None:
                self.logger.warning('Could not persist freshness state: %s', e)

    def flush(self):
        """Force any pending marks to disk (shutdown, or a test)."""
        self._flush(force=True)

    def _load(self):
        path = self.path()
        if not os.path.exists(path):
            return  # cold start: no counter, no marks, first message accepted
        try:
            with open(path) as f:
                doc = json.load(f)
            if not isinstance(doc, dict):
                raise ValueError('not an object')
            seq = int(doc.get('seq', 0))
            marks = {}
            for flat, n in (doc.get('marks') or {}).items():
                sender, _, verb = str(flat).partition('|')
                if not sender or not verb:
                    raise ValueError('malformed mark key %r' % flat)
                marks[(sender, verb)] = int(n)
        except (OSError, IOError, ValueError, TypeError) as e:
            # Refused loudly and left in place rather than silently treated as
            # empty: "no marks" is the state an attacker would want to induce,
            # and an operator needs to see that the file, not the protocol,
            # is what needs fixing.
            if self.logger is not None:
                self.logger.error(
                    'Freshness state unreadable (%s); replay protection for '
                    '%s starts from empty this boot', e, self.proc_name)
            return
        self._seq = seq
        self._marks = marks
        if self.logger is not None:
            self.logger.debug(
                'Freshness restored for %s: seq=%d, %d mark(s)',
                self.proc_name, self._seq, len(self._marks))


def _monotonic():
    import time
    return time.monotonic()

File: src/c/autonomous_trust/utilities/freshness.c
/*******************
 * Copyright 2026 TekFive, Inc., Sean M. Brennan, and contributors
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *******************/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include <jansson.h>

#include "config/configuration.h"
#include "config/discover.h"   /* CFG_FILE_EXT */
#include "structures/data.h"
#include "utilities/exception.h"
#include "utilities/freshness.h"

/* Key form for the marks map. Flat "sender|verb" because jansson has no
 * composite object key and the pair is already two short strings. */
#define FRESHNESS_KEY_LEN 192

static int64_t _fresh_now_us(void)
{
    struct timespec ts;
    if (clock_gettime(CLOCK_MONOTONIC, &ts) != 0)
        return 0;
    return (int64_t)ts.tv_sec * 1000000LL + (int64_t)ts.tv_nsec / 1000LL;
}

static int _fresh_path(const freshness_t *fr, char *out, size_t cap)
{
    char cfg_dir[CFG_PATH_LEN + 1] = {0};
    if (get_cfg_dir(cfg_dir, sizeof(cfg_dir)) < 0)
        return -1;
    int n = snprintf(out, cap, "%s/%s-%s%s", cfg_dir, FRESHNESS_FILE,
                     fr->proc_name, CFG_FILE_EXT);
    if (n < 0 || (size_t)n >= cap)
        return -1;
    return 0;
}

static void _fresh_key(char *out, size_t cap, const char *sender,
                       const char *verb)
{
    snprintf(out, cap, "%s|%s", sender == NULL ? "" : sender,
             verb == NULL ? "" : verb);
}

static void _fresh_write(freshness_t *fr, bool force, logger_t *logger)
{
    if (!force && !fr->dirty)
        return;
    int64_t now_us = _fresh_now_us();
    if (!force && (now_us - fr->last_flush_us) < FRESHNESS_FLUSH_US)
        return;

    char path[CFG_PATH_LEN + 96] = {0};
    if (_fresh_path(fr, path, sizeof(path)) != 0)
    {
        log_warn(logger, "Freshness: cannot build path; not persisting\n");
        return;
    }
    json_t *doc = json_object();
    json_t *marks = json_object();
    if (doc == NULL || marks == NULL)
    {
        if (doc != NULL) json_decref(doc);
        if (marks != NULL) json_decref(marks);
        return;
    }
    json_object_set_new(doc, "seq", json_integer((json_int_t)fr->seq));
    map_key_t key = NULL;
    data_t *val = NULL;
    map_entries_for_each(&fr->marks, key, val)
    {
        int n = 0;
        if (data_integer(val, &n) == 0)
            json_object_set_new(marks, (const char *)key,
                                json_integer((json_int_t)n));
    }
    map_end_for_each
    json_object_set_new(doc, "marks", marks);

    /* Swallowed like the other snapshot writers: the alternative is taking a
     * process down over a full disk. Logged at warning rather than debug
     * because the cost is a reopened replay window. */
    if (json_dump_file(doc, path, JSON_INDENT(2)) != 0)
        log_warn(logger, "Freshness: could not persist to %s\n", path);
    else
    {
        fr->dirty = false;
        fr->last_flush_us = now_us;
    }
    json_decref(doc);
}

static void _fresh_load(freshness_t *fr, logger_t *logger)
{
    char path[CFG_PATH_LEN + 96] = {0};
    if (_fresh_path(fr, path, sizeof(path)) != 0)
        return;
    json_error_t err;
    json_t *doc = json_load_file(path, 0, &err);
    if (doc == NULL)
        return;  /* cold start: no counter, no marks, first message accepted */
    if (!json_is_object(doc))
    {
        log_error(logger,
                  "Freshness: state malformed (%s); replay protection for %s "
                  "starts from empty this boot\n", path, fr->proc_name);
        json_decref(doc);
        return;
    }
    json_t *j_seq = json_object_get(doc, "seq");
    if (json_is_integer(j_seq))
        fr->seq = (int64_t)json_integer_value(j_seq);
    json_t *j_marks = json_object_get(doc, "marks");
    size_t restored = 0;
    if (json_is_object(j_marks))
    {
        const char *k = NULL;
        json_t *v = NULL;
        json_object_foreach(j_marks, k, v)
        {
            if (!json_is_integer(v))
                continue;
            map_set(&fr->marks, (map_key_t)k,
                    integer_data((int)json_integer_value(v)));
            restored++;
        }
    }
    json_decref(doc);
    log_debug(logger, "Freshness restored for %s: seq=%lld, %zu mark(s)\n",
              fr->proc_name, (long long)fr->seq, restored);
}

int freshness_init(freshness_t *fr, const char *proc_name, logger_t *logger)
{
    if (fr == NULL || proc_name == NULL)
        return EXCEPTION(EINVAL);
    memset(fr, 0, sizeof(*fr));
    snprintf(fr->proc_name, sizeof(fr->proc_name), "%s", proc_name);
    map_init(&fr->marks);
    map_init(&fr->refusals);
    fr->initialized = true;
    _fresh_load(fr, logger);
    return 0;
}

void freshness_free(freshness_t *fr, logger_t *logger)
{
    if (fr == NULL || !fr->initialized)
        return;
    _fresh_write(fr, true, logger);
    map_free(&fr->marks);
    map_free(&fr->refusals);
    fr->initialized = false;
}

int64_t freshness_stamp(freshness_t *fr, logger_t *logger)
{
    if (fr == NULL || !fr->initialized)
        return 0;
    fr->seq++;
    /* Forced: the caller is about to put this number on the wire. */
    _fresh_write(fr, true, logger);
    return fr->seq;
}

/* Tally one refusal against @p verb. See freshness_refusals() for why the key
 * is the verb alone and why both refusal shapes share the count. */
static void _fresh_refused(freshness_t *fr, const char *verb)
{
    data_t *cur = NULL;
    int count = 0;
    if (map_get(&fr->refusals, (map_key_t)verb, &cur) == 0 && cur != NULL)
        data_integer(cur, &count);
    map_set(&fr->refusals, (map_key_t)verb, integer_data(count + 1));
    fr->refusals_total++;
}

bool freshness_accept(freshness_t *fr, const char *sender, const char *verb,
                      int64_t seq, logger_t *logger)
{
    if (fr == NULL || !fr->initialized || sender == NULL || verb == NULL)
        return false;  /* no state to tally into */
    if (seq <= 0) {
        _fresh_refused(fr, verb);
        return false;  /* unstamped or stripped; both refused */
    }
    char key[FRESHNESS_KEY_LEN];
    _fresh_key(key, sizeof(key), sender, verb);
    data_t *mark = NULL;
    int applied = 0;
    if (map_get(&fr->marks, (map_key_t)key, &mark) == 0 && mark != NULL)
        data_integer(mark, &applied);
    if (seq <= (int64_t)applied) {
        _fresh_refused(fr, verb);
        return false;
    }
    map_set(&fr->marks, (map_key_t)key, integer_data((int)seq));
    fr->dirty = true;
    _fresh_write(fr, false, logger);
    return true;
}

int64_t freshness_mark(freshness_t *fr, const char *sender, const char *verb)
{
    if (fr == NULL || !fr->initialized || sender == NULL || verb == NULL)
        return 0;
    char key[FRESHNESS_KEY_LEN];
    _fresh_key(key, sizeof(key), sender, verb);
    data_t *mark = NULL;
    int applied = 0;
    if (map_get(&fr->marks, (map_key_t)key, &mark) == 0 && mark != NULL)
        data_integer(mark, &applied);
    return (int64_t)applied;
}

int64_t freshness_refusals(freshness_t *fr, const char *verb)
{
    if (fr == NULL || !fr->initialized)
        return 0;
    if (verb == NULL)
        return fr->refusals_total;
    data_t *cur = NULL;
    int count = 0;
    if (map_get(&fr->refusals, (map_key_t)verb, &cur) == 0 && cur != NULL)
        data_integer(cur, &count);
    return (int64_t)count;
}

void freshness_flush(freshness_t *fr, logger_t *logger)
{
    if (fr == NULL || !fr->initialized)
        return;
    _fresh_write(fr, true, logger);
}

void freshness_reset(freshness_t *fr)
{
    if (fr == NULL || !fr->initialized)
        return;
    /* Deliberately no _fresh_write and no _fresh_load: the point is to end up
     * with the cold-start state, and re-reading the file would restore exactly
     * the marks the caller is trying to be rid of. */
    map_free(&fr->marks);
    map_init(&fr->marks);
    map_free(&fr->refusals);
    map_init(&fr->refusals);
    fr->refusals_total = 0;
    fr->seq = 0;
    fr->dirty = false;
    fr->last_flush_us = 0;
}
